2-Draft-Accepted/RFCxxxx-DSC-for-PowerShell7.md

```diff
@@ -0,0 +1,323 @@
+
+- CIM class name prefixes are removed as they are not needed
+- Instances are represented as an array of the type name
+- The comment at the start is dropped as that information is available within `ConfigurationDocument`
+
+### No longer require MOF schema by switching to improved class-based resources
+
+Although DSC resources can be written as WMI/OMI providers, this requires writing
+C/C++ code and compiling for specific Linux distros and processor architectures.
+Writing PowerShell script DSC resources currently still requires a matching schema
+file making it more complex. Instead, the current plan is to
+[only support PowerShell class-based DSC resources](https://github.com/PowerShell/PowerShell/issues/13731)
+in DSC for PowerShell 7. Whereas previously embedded objects were CIMInstance types,
+they will now be .NET types based on PowerShell classes. Additional reasoning
+and discussion is in the issue linked above.
+
+Making the change from script-based resources to class-based resources could make
+learning DSC a frustrating experience for new users. To resolve this, tooling will
+be needed that scaffolds as much of the class requirements as possible based
+on a .psm1 module file containing traditionally authored Get/Set/Test-TargetResource
+functions. This would also aid in migrating the more than 1000 existing resources
+for Windows.
+
+### Open Source
+
+The current
+[PSDesiredStateConfiguration](https://docs.microsoft.com/powershell/module/psdesiredstateconfiguration/)
+module is not Open Source. The intent is to clean up the code and prepare the GitHub
+repository to be made public as an Open Source project.
+However, DSC for PowerShell 7 does not depend on PSDesiredStateConfiguration being Open Source.
+
+### PowerShell engine changes
+
+To enable DSC for PowerShell 7 to evolve independently to PowerShell 7 itself, the
+code in the engine that handles the DSC configuration keyword will be separated
+as a plugin and made to be part of the PSDesiredStateConfiguration module.
+
+## Alternate Proposals and Considerations
+
+### Supporting existing script based resources
+
+Existing script-based resources (not class-based) require a MOF schema file for
+PowerShell to understand the types and nested types. We had prototyped using a JSON
+based schema instead of the existing MOF schema file but it still required the DSC
+resource author to create a schema file and ensure it aligns with the script implementation.
+
+It is simpler to create a PowerShell class based resource which itself is the schema
+removing the need to understand and maintain a separate schema file. In addition,
+having a single model (class based) for authoring DSC resources will help the community
+in the long term by having a single consistent way to write DSC resources, making
+it easier for new authors to find
+[documentation and examples](https://dsccommunity.org/blog/class-based-dsc-resources/).
+
+### Tooling for parsing JSON configuration
+
```

Comment:
as an aside, we need to be sure that a schema for the JSON is published or otherwise available for a good authoring experience.

Reply:
I see an opportunity for a tool or VSCode plugin to generate a schema on the fly from available resources. Based on the example json earlier in the RFC, the schema will end up very generic and not able to offer more than basic structural validation. An auto generated schema can offer a much stronger validation and auto complete for resource properties.